java-checks/src/main/java/org/sonar/java/checks/spring/SpringScanDefaultPackageCheck.java

```diff
@@ -0,0 +1,171 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.spring;
+
+import com.google.common.collect.ImmutableList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ConstantUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.CompilationUnitTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.NewArrayTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4602")
+public class SpringScanDefaultPackageCheck extends IssuableSubscriptionVisitor {
+
+  private static final String DEFAULT_ATTRIBUTE = "value";
+
+  private static final Map<String, Set<String>> SCAN_PACKAGE_ATTRIBUTES = buildScanPackageAttributes();
+
+  private static Map<String, Set<String>> buildScanPackageAttributes() {
+    Map<String, Set<String>> map = new HashMap<>();
+
+    map.put("org.springframework.context.annotation.ComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    map.put("org.springframework.boot.autoconfigure.SpringBootApplication",
+      new HashSet<>(Arrays.asList("scanBasePackages", "scanBasePackageClasses")));
+
+    map.put("org.springframework.boot.web.servlet.ServletComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    return map;
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.ANNOTATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    AnnotationTree annotation = (AnnotationTree) tree;
+    Set<String> scanPackageAttributeNames = SCAN_PACKAGE_ATTRIBUTES.get(annotation.symbolType().fullyQualifiedName());
+    if (scanPackageAttributeNames != null) {
+      List<ExpressionTree> scanPackageAttributes = annotation.arguments().stream()
+        .filter(argument -> scanPackageAttributeNames.contains(attributeName(argument)))
+        .collect(Collectors.toList());
+
+      if (scanPackageAttributes.isEmpty()) {
+        checkIfAnnotationIsInDefaultPackage(annotation);
+      } else {
+        checkAnnotationPackageAttributes(annotation, scanPackageAttributes);
```

Comment:
IMO you should always pass the scanPackageValues to one method and if it is empty check the default package.

Reply:
I moved the ".isEmpty()" test inside "checkAnnotationPackageAttributes"